File: src/client/java/com/somefrills/features/core/EventSubscriptions.java
package com.somefrills.features.core;

import com.somefrills.Main;
import meteordevelopment.orbit.listeners.IListener;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

public final class EventSubscriptions {
    private EventSubscriptions() {}

    private static final Map<Object, IListener> listeners = new HashMap<>();

    /**
     * Register callback for event type.
     */
    public static <T> void register(Object owner, Class<T> eventClass, Consumer<T> action) {
        unregister(owner);

        IListener listener = new IListener() {
            @Override
            @SuppressWarnings("unchecked")
            public void call(Object event) {
                action.accept((T) event);
            }

            @Override
            public Class<?> getTarget() {
                return eventClass;
            }

            @Override
            public int getPriority() {
                return 0;
            }

            @Override
            @Deprecated
            public boolean isStatic() {
                return false;
            }
        };

        listeners.put(owner, listener);
        Main.eventBus.subscribe(listener);
    }

    /**
     * Convenience overload for no event usage.
     */
    public static <T> void register(Object owner, Class<T> eventClass, Runnable action) {
        register(owner, eventClass, e -> action.run());
    }

    public static <T> void register(AbstractFeature feature, Class<T> eventClass) {
        register(feature, eventClass, e -> feature.sync());
    }

    public static void unregister(Object owner) {
        IListener listener = listeners.remove(owner);
        if (listener != null) {
            Main.eventBus.unsubscribe(listener);
        }
    }

    public static void clear() {
        for (IListener listener : listeners.values()) {
            Main.eventBus.unsubscribe(listener);
        }
        listeners.clear();
    }
}

File: src/main/kotlin/com/somefrills/features/core/EventSubscriptions.kt
package com.somefrills.features.core

import com.somefrills.events.FrillsEvent
import com.somefrills.events.core.EventListeners.Listener
import com.somefrills.events.core.FrillsEvents
import java.util.function.Consumer

object EventSubscriptions {

    data class Subscription(
        val eventClass: Class<Any>,
        val listener: Listener
    )
    private val listeners: MutableMap<Any, Subscription> = HashMap()

    /**
     * Register callback for event type.
     */
    @Suppress("UNCHECKED_CAST")
    fun <T : FrillsEvent> register(owner: Any, eventClass: Class<T>, action: Consumer<T>) {
        unregister(owner)

        val listener = Listener("onEvent", action as Consumer<Any>, 0)

        listeners[owner] = Subscription(eventClass as Class<Any>, listener)
        FrillsEvents.registerListener(eventClass, listener)
    }

    /**
     * Convenience overload for no event usage.
     */
    fun <T : FrillsEvent> register(owner: Any, eventClass: Class<T>, action: Runnable) {
        register(owner, eventClass) { e: T -> action.run() }
    }

    fun <T : FrillsEvent> register(feature: AbstractFeature, eventClass: Class<T>) {
        register(feature, eventClass) { e: T -> feature.sync() }
    }

    @Suppress("UNCHECKED_CAST")
    fun unregister(owner: Any) {
        val listener = listeners.remove(owner)
        if (listener != null) {
            FrillsEvents.unregisterListener(listener.eventClass as Class<out FrillsEvent>, listener.listener)
        }
    }

    fun clear() {
        val owners = listeners.keys.toList()
        for (owner in owners) {
            unregister(owner)
        }
        listeners.clear()
    }
}